### PR TITLE
[Frontendv1] Make the execution time human readable on UI.

### LINKED ISSUE #3319: Make the execution time human readable on UI.
**Issue**:
We now show the `Execution time` on the leaderboard but it is in seconds. We would like to make it more human-readable on UI like for `3600` seconds, it should display 1 hour.
For e.g. in the below screenshot, `134983.881963 seconds` isn't readable.

image

**Tasks:**
- [ ] Convert the seconds from the API to human-readable format in the controller.

Please feel free to ask if you have any doubts/questions.

Note: This is a high very priority issue, please pick it up only when a PR can be opened within 24 hours.

### PR DESCRIPTION
**Pull request by Alabhya268**
fixes #3319
- [x] Make the execution time human readable on UI.

I have made execution time human readable.
I have not yet add ```day``` as I think it unecessay but let me know if its needed to be added.

time

### CODE REVIEW

**Comment by Alabhya268** [comment]
@RishabhJain2018  @Ram81 , Please review this PR.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 1]
```diff
@@ -872,6 +872,14 @@
             scope.sortColumn = column;
         };
 
+        vm.format_execution_time = function(execution_time) {
+            var execution_time_hr = Math.floor(execution_time/60/60);
+            var execution_time_min = Math.floor(execution_time/60) - (execution_time_hr * 60);
+            var execution_time_sec = Math.floor(execution_time % 60);
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:878]: @Alabhya268 can we use date util to do the conversion. See [this](https://stackoverflow.com/questions/1322732/convert-seconds-to-hh-mm-ss-with-javascript) for reference

File: frontend/src/js/controllers/challengeCtrl.js [hunk 2]
```diff
@@ -872,6 +872,14 @@
...
+            var formatted = execution_time_hr + 'hr ' + execution_time_min + 'min ' + execution_time_sec + 'sec';
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:879]: Add space before `Hr` `Min` and `Sec`

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Comment by RishabhJain2018** [comment]
>I have not yet add day as I think it unecessay but let me know if its needed to be added.

Also, please do add day as some submissions run for over 36 hours.

File: frontend/src/js/controllers/challengeCtrl.js (hunk 1 above)
Comment (Alabhya268) [reply frontend/src/js/controllers/challengeCtrl.js:878]: @Ram81, I tried to do it in a way . It not working as expected.
Its adding the executions time in current time. I tried to modify it , I am able to get hour, min and sec but not days with this method.

File: frontend/src/js/controllers/challengeCtrl.js (hunk 2 above)
Comment (Alabhya268) [reply frontend/src/js/controllers/challengeCtrl.js:879]: ok

**Comment by Alabhya268** [comment]
@Ram81, Please have a look at this PR, I have now done it with date util, but for some reason date was always 1 day ahead so I added ```-1``` , Could you please test it ?

@RishabhJain2018 , I have now added ```day``` option.
time2

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
@Alabhya268 Can we please create a filter like this: https://stackoverflow.com/a/30767487 which we can use at multiple places i.e. in all submissions and my submissions as well.

**Comment by Alabhya268** [comment]
> @Alabhya268 Can we please create a filter like this: https://stackoverflow.com/a/30767487 which we can use at multiple places i.e. in all submissions and my submissions as well.

I will try to do it .

**Comment by Alabhya268** [comment]
> @Alabhya268 Can we please create a filter like this: https://stackoverflow.com/a/30767487 which we can use at multiple places i.e. in all submissions and my submissions as well.

@RishabhJain2018 , I have made it as filter and it can now be used anywhere to convert seconds to datetime.

**Comment by Alabhya268** [comment]
> @Alabhya268 Can we please create a filter like this: https://stackoverflow.com/a/30767487 which we can use at multiple places i.e. in all submissions and my submissions as well.

@RishabhJain2018 , I have made it as filter and it can now be used anywhere to convert seconds to datetime.

**Comment by RishabhJain2018** [comment]
@Alabhya268 Can you please fix the Travis build?

**Comment by Alabhya268** [comment]
@RishabhJain2018 Travis build is now fixed.

File: frontend/src/js/filters/filters.js [hunk 3]
```diff
@@ -15,4 +15,21 @@
...
+            .filter('format_execution_time', format_execution_time);
+
+    function format_execution_time() {
+        return function(execution_time) {
+            var executiontime = new Date(execution_time * 1000);
+            var days = executiontime.getUTCDate() - 1;
+            var hours = executiontime.getUTCHours();
+            var minutes = executiontime.getUTCMinutes();
+            var seconds = executiontime.getSeconds();
+            var timeString = days.toString().padStart(2, '0') + ' day ' + hours.toString().padStart(2, '0') + ' hr ' +
```
Comment (Ram81) [inline frontend/src/js/filters/filters.js:28]: @Alabhya268 let's not show `day` `'hr` or `min` when they have 0 value. Rest of the changes look fine.

cc: @RishabhJain2018

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Comment by Alabhya268** [comment]
@Ram81 , days , min , hours will now show accordingly.

**Comment by Alabhya268** [comment]
@Ram81 . Ping for PR

**Comment by Alabhya268** [comment]
@Ram81, Please review this PR again.

**Comment by Ram81** [comment]
@Alabhya268 If a submission takes 1 day we are showing 1 day and 00 sec we should either show all hour, mins and secs if we are showing day or we should only show when value is greater than 0.

Screen Shot 2021-03-21 at 10 04 04 PM

@RishabhJain2018 what should we do? should we just stick to showing hh : mm : ss even when submission takes more than 24 hrs?

**Comment by Alabhya268** [comment]
> @Alabhya268 If a submission takes 1 day we are showing 1 day and 00 sec we should either show all hour, mins and secs if we are showing day or we should only show when value is greater than 0.
> 
> Screen Shot 2021-03-21 at 10 04 04 PM
> 
> @RishabhJain2018 what should we do? should we just stick to showing hh : mm : ss even when submission takes more than 24 hrs?

@Ram81 , I think we should stick full ``` hh : mm  : ss```  as UI would look odd if different challenges in column uses different format to show execution time. 

CC: @RishabhJain2018

**Comment by Alabhya268** [comment]
> @Alabhya268 If a submission takes 1 day we are showing 1 day and 00 sec we should either show all hour, mins and secs if we are showing day or we should only show when value is greater than 0.
> 
> Screen Shot 2021-03-21 at 10 04 04 PM
> 
> @RishabhJain2018 what should we do? should we just stick to showing hh : mm : ss even when submission takes more than 24 hrs?

@Ram81 @RishabhJain2018 , Which approch should I go for ?

**Comment by Ram81** [comment]
@Alabhya268 let's just show day hh mm sec when the value is > 0. And just show sec if submission time is > a minute. Let me know if you have any doubts

**Comment by Alabhya268** [comment]
> @Alabhya268 let's just show day hh mm sec when the value is > 0. And just show sec if submission time is > a minute. Let me know if you have any doubts

@Ram81 , It means we just need to show everthing till highest time right ?
like if execution is 1 hour , we'll show ```hh:mm:ss``` .

**Comment by Alabhya268** [comment]
> > @Alabhya268 let's just show day hh mm sec when the value is > 0. And just show sec if submission time is > a minute. Let me know if you have any doubts
> 
> @Ram81 , It means we just need to show everthing till highest time right ?
> like if execution is 1 hour , we'll show `hh:mm:ss` .

@Ram81 ?

**Comment by Alabhya268** [comment]
> > @Alabhya268 let's just show day hh mm sec when the value is > 0. And just show sec if submission time is > a minute. Let me know if you have any doubts
> 
> @Ram81 , It means we just need to show everthing till highest time right ?
> like if execution is 1 hour , we'll show `hh:mm:ss` .

@Ram81 , @RishabhJain2018

**Comment by Alabhya268** [comment]
> > @Alabhya268 let's just show day hh mm sec when the value is > 0. And just show sec if submission time is > a minute. Let me know if you have any doubts
> 
> @Ram81, you mean we show sec when time is < min right?
> 
> @Ram81, I am gonna make it so that `dd: hh: mm: ss` always show when execution time is > 0 second, This would make the execution time of the challenge list look even.

@RishabhJain2018, Please also state your thoughts.

**Comment by Alabhya268** [comment]
@Ram81, I've made it so that it will show ```dd: hh: mm: ss``` when execution time is > 0 and it will show ```00 sec``` when execution time is <1.

**Comment by Ram81** [comment]
@Alabhya268 can you share a screenshot with multiple submission each having different execution time. For ex: 10 sec, 1 min, 1 min 30 sec, 1 hour, 1 hour 20 min, 1 hour 30 sec, 1 hour 20 min 30 sec, 1 day, 1 day 10 sec, etc

**Comment by Alabhya268** [comment]
> @Alabhya268 can you share a screenshot with multiple submission each having different execution time. For ex: 10 sec, 1 min, 1 min 30 sec, 1 hour, 1 hour 20 min, 1 hour 30 sec, 1 hour 20 min 30 sec, 1 day, 1 day 10 sec, etc
@Ram81, Please look at these screenshot. If you have something different in mind then please let me know precisely what format should look.

1hour 20 min
1hour 30sec
1min
1min 30 sec
10sec
s
Screenshot from 2021-03-27 11-53-03
1day
1day 1 mon 12 sec
1day 10 sec
1h 20 min 30 sec
1hour

**Comment by Ram81** [comment]
@Alabhya268 I think I didn't clearly mention it. We want to show `days`, `min`, `hr` only if the value is > 0. Here are the expected values for few cases:

24 hr 10 sec -> 01 day 10 secs
90 secs -> 1 min 30 secs
26 hr 2mins -> 1 day 122 mins

Let me know if you have any doubts

**Comment by Alabhya268** [comment]
> @Alabhya268 I think I didn't clearly mention it. We want to show `days`, `min`, `hr` only if the value is > 0. Here are the expected values for few cases:
> 
> 24 hr 10 sec -> 01 day 10 secs
> 90 secs -> 1 min 30 secs
> 26 hr 2mins -> 1 day 122 mins
> 
> Let me know if you have any doubts

@Ram81, I will incorporate the requested changes asap.

**Comment by Alabhya268** [comment]
> > @Alabhya268 I think I didn't clearly mention it. We want to show `days`, `min`, `hr` only if the value is > 0. Here are the expected values for few cases:
> > 24 hr 10 sec -> 01 day 10 secs
> > 90 secs -> 1 min 30 secs
> > 26 hr 2mins -> 1 day 122 mins
> > Let me know if you have any doubts
> 
> @Ram81, I will incorporate the requested changes asap.

@Ram81, Wouldn't we want to show 26 hours 2 min -> 1 day 2 hour 2 mins?

**Comment by Ram81** [comment]
Yes, updated the previous comment

**Comment by Alabhya268** [comment]
@Ram81, I've made execution time according to your requirements. Please have a look at the following images. 

Screenshot from 2021-03-28 17-53-01
Screenshot from 2021-03-28 17-53-38
Screenshot from 2021-03-28 18-07-12
Screenshot from 2021-03-28 18-07-28

**Comment by Alabhya268** [comment]
@Ram81 ping for review

**Comment by Alabhya268** [comment]
> @Ram81, I've made execution time according to your requirements. Please have a look at the following images.
> 
> Screenshot from 2021-03-28 17-53-01
> Screenshot from 2021-03-28 17-53-38
> Screenshot from 2021-03-28 18-07-12
> Screenshot from 2021-03-28 18-07-28

@Ram81 , Please have a look at this PR, I've made execution time according to your requirements.

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**